### ISSUE TITLE
Conda repo as a channel

### ISSUE THREAD

**Issue by marctorsoc** [python-poetry/poetry#1904]
- [ ] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [ ] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

(just copying from my comment in the feature request thread. Apologies for the spam there.
I understand this is something some people would find useful, and even though the maintainer already expressed this is not in current plans, I just opened the issue so that we can collect more views. Therefore, I would ask to keep it open for some months to grasp whether this is a desired feature or not)

Something I would like to see in future releases is the ability to use different channels. In conda you can specify e.g.

name: dslib
channels:
  - anaconda
  - conda-forge
dependencies:
  # python related packages
  - python==3.6.5
  - scikit-learn 
However, you cannot install anything outside of conda, which is really bad. That's why (among other things) I and other people I know use poetry.

On the other hand, poetry only supports pip, which is not perfect as explained here https://www.anaconda.com/understanding-conda-and-pip/

TLDR:

There is never a need to have compilers available to install them. You download binaries directly, so installation is cleaner and faster
Additionally conda packages are not limited to Python software.
Is there any chance to get support for conda? I would like to see something like channels, and poetry checking first conda, then pip. This would probably improve installation speed as well

**Comment by wimvanleuven-kbc** [comment]
I would love to see poetry support anaconda as well. Similar to virtual envs for Python, Anaconda is though much better at handling version (in)compatibiliities. At the cost though of being slower to create an environment from scratch because of all the dependency resolutions happening.

**Comment by abn** [comment]
@marctorsoc at initial glance, I feel this is out of scope for poetry. Also, if anaconda already does version resolution, wouldn't that be redundant effort since poetry already does this? All that said, it could be a future third-party plugin feature if you really want something like this.

I think it could be helpful if you detail how this would work if implemented. An example project with a `pyproject.toml`, perhaps detailing what commands like `add`, `install`, `run` and `shell` would be expected to do etc.

**Comment by finswimmer** [comment]
I don't think, supporting conda channels for python packages doesn't add any benefit to poetry. Usually python packages available in a conda repository are also available on PyPa.

A really great benefit, particularly if you are doing datascience or bioinformatics, would be to support non-python applications from the conda channels. But - at least for now - this is definitely out of scope for poetry, as this would add a complete different ecosystem to the dependency solver.

If  there is someone out there with a lot of time .... :)

fin swimmer

**Comment by marctorsoc** [comment]
> @marctorsoc at initial glance, I feel this is out of scope for poetry. Also, if anaconda already does version resolution, wouldn't that be redundant effort since poetry already does this? All that said, it could be a future third-party plugin feature if you really want something like this.
> 
> I think it could be helpful if you detail how this would work if implemented. An example project with a `pyproject.toml`, perhaps detailing what commands like `add`, `install`, `run` and `shell` would be expected to do etc.

With all respect @abn  , I don't think you fully read my initial post where I mentioned the advantages of using conda vs pip, one of them pointed out my @finswimmer . To reiterate:

* conda packages are not limited to Python software
* no need of compilers
* faster installation than pip

As per why use poetry then if I have conda:
* Because poetry allows for easy  reproduction of the environment with the `pyproject.toml`, where `poetry add` adds only the package you mention but not all dependencies. I don't think there's anything equivalent for conda
* Some packages are not in conda
* Not easy to install my own packages using conda

**Comment by abn** [comment]
@marctorsoc I did read your initial post ;) We should clarify a few things here.

The conversation here should not necessarily be about `pip` vs `conda`, at least not from `poetry`'s perspective. It is important to note that the dependency on `pip` might be dropped or reduced in a future release but this is still an ongoing conversation. We should instead be talking about how `poetry`'s internal architecture could evolve to enable plugins that could implement such use cases. As things stand today, implementing something like this will require significant changes as @finswimmer has already pointed out. I would much rather these changes be more generic than being specific to `conda`.

The cross-platform nature, availablity of pre-compiled binaries (platform & python packages) and distribution of compilers via `conda` itself are advantages. However, this does not necessarily make it faster or desired in all cases.

While `conda` can be used within a standard virtualenv, in order to make full use of all its features (including distributed compilers, pre-compiled binaries etc.), it needs to be used within an environment managed by `conda` itself. Mixing and matching `pip` and `conda` in the same environment might be less than ideal. And as far as I can tell, if a binary was compiled for/in a conda managed environment, there is no guarantee that it will work in a standard python environment created by `virtualenv` or `venv`.

Considering the above, it seems to me that enabling this will require `poetry` to be aware that the project environment is being managed by `conda`. This will most likely mean that `poetry` will have to somehow offload managing environments themselves and packages to `conda`, falling back to PyPI if package was not found. Additionally, it will need to retrive package metadata, atleast for python packages, from `conda` or the corresponding `channel` directly (similar to how PyPI repo is handled) via a standard interface. I am going to guess publishing such packages will need to have certain considerations as well.

In any case, this is merely my 2c on how to approach this, hopefully this is useful to help drive the conversation forward. Once there is a concrete plan, we can definitely talk what changes are required within `poetry` itself to enable it.

As I said before, if this is really required, this is probably better off and cleaner to implement as its own plugin to `poetry`. I suspect we will have to make changes to `poetry`'s internals in ordeer to enable such a use case. As far as `poetry` itself is concerned, as @finswimmer and myself have indicated earlier, this is definitely out of scope.

**Comment by marctorsoc** [comment]
Thanks for your answer @abn . I always use `conda` as my virtualenv manager, but I appreciate that's not always the case, especially for out-of-data-science folks. 

I have used poetry with a conda environment and was working fine. We basically were creating the env with conda with some initial packages to exploit mkl, and everything else would go via pip with poetry. 

Hope some day poetry will support this, or someone in anaconda will make the bridge easier to cross

**Comment by wimvanleuven-kbc** [comment]
If I understand it correctly, Poetry already supports the concept of Envs and an EnvManager through which all commands to manage environments and dependencies are routed. Correct? 

So adding Conda support would require a CondaEnvManager and CondaEnvs of somesort? 

What could become an issue is that managing dependencies 1 by 1 might be slow to Conda, so adding and updating an environment in bulk will be required here. Is that something that is aligned with / exists in Poetry's current design?

**Comment by sknutsonsf** [comment]
Better interaction between poetry and conda would be appreciated.  Currently I install a basic environment via conda then use pip to get everything else I need.  This two-phase building of the environment causes headaches and risks of incompatible versions.  It would be nice to build my environment with just one tool.  I really need Conda basic for numpy, scipy and numba, all of which are difficult to install without conda.
